custom/enikshay/tasks.py

```diff
@@ -0,0 +1,296 @@
+        1. Find most relevant case
+            if only non-enikshay source cases
+                consider the case with latest_modified - irrespective of case is closed/open
+
+            if only enikshay source cases
+                filter by '(closed and closure_reason == HISTORICAL_CLOSURE_REASON) or open'
+                consider the case with latest modified after above filter
+
+            if mix of enikshay and non-enikshay source cases
+                ignore non-enikshay and apply above enikshay only condition
+        2. Check if 'adherence_value' of most relevent case is one of DOSE_TAKEN_INDICATORS
+        """
+        def is_dose_taken(cases):
+            # runs above discribed calculation and returns whether a dose is taken or not
+            sources = set(map(lambda x: x["adherence_source"], cases))
+            if 'enikshay' not in sources:
+                valid_cases = cases
+            else:
+                valid_cases = filter(
+                    lambda case: (
+                        case['adherence_source'] is 'enikshay' and
+                        (not case['closed'] or (case['closed'] and
+                         case['adherence_closure_reason'] == HISTORICAL_CLOSURE_REASON))
+                    ),
+                    cases
+                )
+            if valid_cases:
+                by_modified_on = sorted(valid_cases, key=lambda case: case['modified_on'])
+                return by_modified_on[-1]['adherence_value'] in DOSE_TAKEN_INDICATORS
+            else:
+                return False
+
+        # index by 'adherence_date'
+        cases_by_date = defaultdict(list)
+        for case in adherence_cases:
+            adherence_date = parse_datetime(case['adherence_date']).date()
+            cases_by_date[adherence_date].append(case)
+
+        # calculate whether adherence is taken on each day
+        dose_taken_by_date = defaultdict(bool)
+        for d, cases in cases_by_date.iteritems():
+            dose_taken_by_date[d] = is_dose_taken(cases)
+
+        return dose_taken_by_date
+
+    def count_doses_taken(self, adherence_cases, lte=None, gte=None):
+        """
+        Args:
+            adherence_cases: list of 'adherence_cases' with its 'adherence_value' as one of DOSE_KNOWN_INDICATORS
+
+        Returns:
+            total count of adherence_cases excluding duplicates on a given day. If there are
+            two adherence_cases on one day at different time, it will be counted as one
+        """
+
+        if not self._cache_dose_taken_by_date and self._cache_dose_taken_by_date != {}:
+            # memoize this, `calculate_doses_taken_by_day` can't be memoized using decorator as it uses list arg
+            dose_taken_by_date = self.calculate_doses_taken_by_day(adherence_cases)
+            self._cache_dose_taken_by_date = dose_taken_by_date
+        else:
+            dose_taken_by_date = self._cache_dose_taken_by_date
```

Comment:
Hey, I just realized that you can probably do away with this memoization dict (which sort of obfuscates what's going on) by just passing `dose_taken_by_date` to `count_doses_taken()` instead of `adherence_cases`. I think this change would be nice to have, but not blocking the PR since I know this is on a tight timeline.

e.g.
```python
adherence_cases = self.get_valid_adherence_cases()
dose_taken_by_date = self.calculate_doses_taken_by_day(adherence_cases)
update["adherence_total_doses_taken"] = self.count_doses_taken(dose_taken_by_date)
# calculate 'aggregated_score_count_taken'
update["aggregated_score_count_taken"] = self.count_doses_taken(
	dose_taken_by_date,
	lte=adherence_schedule_date_start,
	gte=update["aggregated_score_date_calculated"]
)
```